Optionally disable "render back faces" for the shadow caster

WilliamLewww opened the pull request:
Added a new SDF parameter `ignition:shadow_caster_render_back_faces` to specify if back faces should be rendered by the shadow caster.

```
<scene>
  <ignition:shadow_caster_render_back_faces>0</ignition:shadow_caster_render_back_faces>
  <background>0.0 0.0 0.0 1</background>
</scene>
```

Currently, the scene is hard-coded to always render back faces.


Based off pull request: #3048

WilliamLewww committed:
Added SDF parameter for render back faces

WilliamLewww committed:
Added SDF parameter for render back faces

iche033 commented on gazebo/physics/WorldPrivate.hh:
```diff
@@ -391,6 +391,9 @@ namespace gazebo
       public: std::string shadowCasterMaterialName;
+
+      /// \brief Shadow caster render back faces from scene SDF
+      public: bool shadowCasterRenderBackFaces;
```
set default value here.

iche033 commented on gazebo/physics/World.cc:
```diff
@@ -229,6 +229,18 @@ void World::Load(sdf::ElementPtr _sdf)
+  }
+  else
+  {
+    this->dataPtr->shadowCasterRenderBackFaces = 1;
```
can we do `true|false` here? Easier to find out the underlying type. Same for the one in ScenePrivate.hh

iche033 commented on gazebo/physics/World.cc:
```diff
@@ -229,6 +229,18 @@ void World::Load(sdf::ElementPtr _sdf)
+    this->dataPtr->shadowCasterRenderBackFaces =
+      this->dataPtr->sdf->GetElement("scene")->
+        Get<bool>("ignition:shadow_caster_render_back_faces");
+  }
```
this looks fine. I think if we plane to add more shadow params in the future, we can consider adding a `<shadows> </shadows>` section to group these parameters

iche033 reviewed (commented):
looks good, just some minor comments

scpeters commented:
we could consider adding this new field to scene.proto in the ign-msgs `main` branch, similar to https://github.com/ignitionrobotics/ign-msgs/pull/179

WilliamLewww committed:
Added default value for shaderCasterRenderBackFacesr

Signed-off-by: William Lew <WilliamMilesLew@gmail.com>

iche033 commented on gazebo/rendering/ScenePrivate.hh:
```diff
@@ -379,6 +379,9 @@ namespace gazebo
       public: std::string shadowCasterMaterialName = "Gazebo/shadow_caster";
+
+      /// \brief Shadow caster render back faces
+      public: bool shadowCasterRenderBackFaces = 1;
```
can you set this to `= true` as well?

iche033 reviewed (approved):
just made one very minor comment. Otherwise looks good.

WilliamLewww committed:
Use true/false for initializing bool

Signed-off-by: William Lew <WilliamMilesLew@gmail.com>

scpeters commented on gazebo/physics/World.cc:
```diff
@@ -229,6 +229,18 @@ void World::Load(sdf::ElementPtr _sdf)
+  else
+  {
+    this->dataPtr->shadowCasterRenderBackFaces = true;
+  }
```
the default value is already specified in `WorldPrivate`, so I think this `else` block is unnecessary

scpeters reviewed (commented):
we included a test in #3048. would it be hard to do so here as well?

WilliamLewww committed:
Removed reinitialization of shadowCasterRenderBackFaces and shadowCasterMaterialName

Signed-off-by: William Lew <WilliamMilesLew@gmail.com>

WilliamLewww committed:
Added tests for shadow_caster_render_back_faces param

Signed-off-by: William Lew <WilliamMilesLew@gmail.com>

WilliamLewww commented:
> we included a test in #3048. would it be hard to do so here as well?

I added a test for the `shadow caster render back faces`.

scpeters committed:
fix xml in default_shadow_caster.world

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
> > we included a test in #3048. would it be hard to do so here as well?
> 
> I added a test for the `shadow caster render back faces`.

thanks! the test looks great. I just fixed one more xml glitch in ef2568920b2e9dabb52075fdb41eb9c66259854f and will merge once CI has finished again

scpeters commented:
> > > we included a test in #3048. would it be hard to do so here as well?
> > 
> > 
> > I added a test for the `shadow caster render back faces`.
> 
> thanks! the test looks great. I just fixed one more xml glitch in [ef25689](https://github.com/osrf/gazebo/commit/ef2568920b2e9dabb52075fdb41eb9c66259854f) and will merge once CI has finished again

unfortunately the new test is failing on macOS

* https://build.osrfoundation.org/job/gazebo-ci-pr_any-homebrew-amd64/2574/testReport/junit/(root)/ShadowCasterRenderBackFacesTest/ShadowCasterBackFaces/

I'll try to reproduce and debug the failure on my laptop

WilliamLewww commented:
> > > > we included a test in #3048. would it be hard to do so here as well?
> > > 
> > > 
> > > I added a test for the `shadow caster render back faces`.
> > 
> > 
> > thanks! the test looks great. I just fixed one more xml glitch in [ef25689](https://github.com/osrf/gazebo/commit/ef2568920b2e9dabb52075fdb41eb9c66259854f) and will merge once CI has finished again
> 
> unfortunately the new test is failing on macOS
> 
>     * https://build.osrfoundation.org/job/gazebo-ci-pr_any-homebrew-amd64/2574/testReport/junit/(root)/ShadowCasterRenderBackFacesTest/ShadowCasterBackFaces/
> 
> 
> I'll try to reproduce and debug the failure on my laptop

Could this be the same thing that made the test fail for #3048? I remember resorting to excluding the test on MacOS.

scpeters commented:
the test passes on my laptop; let's just merge it as is
